### ISSUE TITLE
torch backend: block contractions are not batched (61k matmuls, 82M calls per 6x6 realization)

### ISSUE THREAD

**Issue by Ryo-wtnb11** [Ryo-wtnb11/symtenet#353]
## Summary

On the torch backend, block-sparse contractions are not batched: `lower_plan` walks its terms one at a time because `batch_plan` is gated to NumPy, and `compose_lowered` issues one `torch.matmul` per coupled sector. Realizing a 6x6 SU(2) PEPS (P=3, D=32, BP-gauged, untied) makes **82 million Python function calls and 61,058 separate `torch.matmul` calls** for roughly 10^9 flops of arithmetic. Wall time is 107 s on an idle node and 370 s under CPU contention; the arithmetic alone would be under a second.

## Where

`src/tenet/map_view.py`, `lower_plan`, around line 1067:

```python
groups, loose = (
    batch_plan(structure, perm, normalized)
    if ar.infer_backend(t.data[0]) == "numpy"
    else ((), normalized)
)
```

The comment above it says the trade of Python iterations for array operations "has been measured only on NumPy". On torch every term goes through the per-term loop: one `reshape`, one `scaled`, one `add` each.

`src/tenet/ops/map.py`, `compose_lowered`, around line 335:

```python
matmul = lib_fn(backend, "matmul")
for k, i, j in plan.products:
    mats[k] = matmul(da[i], db[j])
```

One matmul per coupled sector, never a batched `bmm` over sectors of equal shape.

## Measurement

cProfile of one `realize` (6x6, P=3, D=32, BP gauged, untied, torch CPU, `OMP_NUM_THREADS=20`, `taskset` to 20 cores), sorted by own time:

```
 ncalls   tottime  function
  61058    27.9 s  torch.matmul                    (0.46 ms per call)
  67996    28.3 s  map_view.py:950 lower_plan
6707746    19.0 s  Tensor.reshape
3322574    19.5 s  map_view.py:892 scaled
3337796     9.3 s  torch.add
1615438    10.3 s  torch.multiply
1685564    11.2 s  map_view.py:1093 cell
              159 s total under the profiler, 82,977,045 calls
```

The same realization with the angles as `requires_grad=True` leaves: 370 s under the profiler against 391 s without, 44M calls against 82M, `torch.matmul` 61,058 calls both times. So this is not the autograd fallback (`lower_plan` declining because torch refuses `out=` under a graph); the fast route is being taken and is itself un-batched on torch.

Under contention from four concurrent GPU jobs on the same node the 61,058 matmuls took 253 s -- 4 ms each -- which is what launch-bound small matmuls look like.

Per-layer timing of the same realization (12 layers, BP before each):

```
layers 1-4     0.4 -> 1.6 s
layers 5-8     3.0 -> 13.9 s
layers 9-12   20.4, 19.0, 23.8, 19.0 s    BP is 70-80% of each
```

BP converges in 1-5 iterations, so the cost is per-iteration, and a BP iteration is ~60 directed-edge contractions -- each of which is a `compose_lowered` over ~100 coupled sectors, hence the matmul count.

## Why it matters

For a VMC optimisation this realization runs once per step and is the only CPU-side piece; with the GPU-side amplitude and score evaluation at 34 s per step across four devices, the un-batched realization is 76% of the step. Measurements (no gradient) pay it once and never notice; optimisation pays it every step.

## What would fix it

1. Extend `batch_plan` to torch. Its grouping is a pure function of the plan and already cached with it; the NumPy path gathers a bucket with `stacked[take]`, scales, sums and writes into the destination with `out=`. Torch has no `out=` under autograd, so the write-back would have to be out of place -- the docstring notes this costs a temporary the size of the bucket -- but out of place is still one array operation per bucket rather than one per term.
2. In `compose_lowered`, group `plan.products` by `(shape(da[i]), shape(db[j]))` and issue `torch.bmm` per group. The grouping is again a pure function of the two structures and can live in `compose_plan`.

Either alone should collapse the 6.7M `reshape` / 3.3M `scaled` / 3.3M `add` counts by two to three orders of magnitude, which is what the NumPy docstring reports (296,953 terms in 2,457 dispatches).

## Reproduce

On a checkout with `circuit_synthesis` available (the realization code is `run.realize` in `docs/benchmarks/circuit_fidelity_20260905/data/run.py` of the PEPS_VMC repo, branch `feat/block-fit`):

```python
import cProfile, pstats
from run import layer_angles, realize
shape = (6, 6)
angles = layer_angles("l6_angles_P3.npz", 3, shape)   # any layer-shared P=3 angles
p = cProfile.Profile(); p.enable()
realize(angles, shape, max_bond=32, max_bond_xe=32, gauged=True, tied=False)
p.disable(); pstats.Stats(p).sort_stats("tottime").print_stats(10)
```

Any SU(2) PEPS at comparable bond and block count will show the same shape: matmul count in the tens of thousands, reshape count in the millions.
